students/john_navitsky/session06/mailroom.py

```diff
@@ -200,62 +253,51 @@ def thank_you_entry():
+        if not selection:
             continue
 
         # parse the string into name components
         current_donor=parse_name(selection)
 
-        # determine if the provided name matches an existing record
-        donor_id=None
-        for donor_index, existing_donor in enumerate(donors):
-            # assemble full name from first_name, last_name
-            existing_full_name = existing_donor["first_name"]+" "+existing_donor["last_name"]
-            # if we get a match, store the index of the donor record
-            if existing_full_name.lower().strip() == current_donor["full_name"].lower().strip():
-                donor_id=donor_index
+        # reject blatantly bad input
+        if current_donor["first_name"] == "" or current_donor["last_name"] == "":
+            print("You must enter both a first and last name.")
+            continue
 
         # prompt for new donation, cancel if None returned
-        new_donation=get_donation_amount(current_donor["informal_name"])
+        new_donation=get_donation_amount(current_donor)
         if new_donation is None:
             print_lines()
             print("Donation cancelled!")
             return
 
-        # donor_id will be None if it didn't match an existing entry
-
-        # add donation to an existing donor
-        if donor_id is not None:
-            donors[donor_id]["donations"].append({ "amount": new_donation})
-            hint="returning"
-            #print("existing donor:",donors[donor_id])
-
-        # add a new donor
-        else:
-            donors.append( { "last_name": ", ".join(filter(None,[current_donor["last_name"], current_donor["suffix"]])),
-                "first_name": current_donor["first_name"],
-                "donations": [ { "amount": new_donation} ] } )
-            # set the donor_id to the new donor
-            donor_id=len(donors)-1
-            hint="new"
-            #print(donors[donor_id])
+        # update the donor list with the new donation
+        hint = update_donor(current_donor, donors, new_donation)
 
         # thank the donor for the new donation
-        print_thank_you(donor_id,hint)
+        print_thank_you(current_donor,hint)
 
 
-def thank_all_donors():
-    for index in range(len(donors)):
-        #dest=sys.stdout
-        donor_name = donors[index]["last_name"]+"_"+donors[index]["first_name"]
-        donor_name = donor_name.strip().lower().replace(" ", "_").replace(",", "")
-        donor_file="thank_you_" + donor_name
-        dest = open(donor_file, "w")
-        print_thank_you(index,"wonderful",dest)
-        dest.close()
+def thank_all_donors(donors,dest_override=None):
+    """ loop through donors, open file, print a thank you letter. """
+    for donor_id in donors:
+        donor=donors[donor_id]
+        file_name="_".join(filter( None, [ "thank_you", donor["last_name"], donor["suffix"], donor["first_name"] ])).lower()
+        file_name=file_name.replace(" ", "_")
+        if not dest_override:
```

Comment:
better to check for None explicitly.

`if dest_override is None:`

Reply:
As a general application of "Explicit is better than implicit.", or is there a specific pitfall?  The reason I ask is I tend to really like the ability to treat objects as a Boolean.

Reply:
None is kind of a special case — I like objects as Boolean too, but None is one thing, rather than a value that happens to be a Falsey value.

If you don’t test for None, then 0 or an empty string or and empty sequence will also be false.